### PR TITLE
Evolv75

### PR DESCRIPTION
**Pull request by Gondolindrim**
Evolv75 support

## Types of Changes
Adds support for the Evolv75 PCB.

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/evolv/config.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+#pragma once
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED // AP for AcheronProject
+#define PRODUCT_ID      0x0075 // AR for Arctic
+#define DEVICE_VER      0x0001 // Revision pre-Alpha
+#define MANUFACTURER    NathanAlpha
+#define PRODUCT         Evolv75
+#define DESCRIPTION	A 75 percent keyboard that can hold a truck
```
Comment (drashna) [inline keyboards/evolv/config.h:26]:
```suggestion
```

File: keyboards/evolv/evolv.c [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "evolv.h"
+
+void encoder_update_user(uint8_t index, bool clockwise) {
```
Comment (drashna) [inline keyboards/evolv/evolv.c:20]:
```suggestion
__attribute__((weak)) void encoder_update_user(uint8_t index, bool clockwise) {
```

File: keyboards/evolv/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#define ____ KC_NO
```
Comment (drashna) [inline keyboards/evolv/keymaps/default/keymap.c:17]: These defines aren't necessary, as these are actually defined in the core.  You can safely remove them. 

https://github.com/qmk/qmk_firmware/blob/a0309db983150fd0197eb620a6ba552d90d29f93/quantum/quantum_keycodes.h#L30-L31


```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/evolv/keymaps/default/keymap.c (hunk 3 above)
Comment (Gondolindrim) [reply keyboards/evolv/keymaps/default/keymap.c]: Committed in eec4bff

File: keyboards/evolv/config.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED // AP for AcheronProject
```
Comment (drashna) [inline keyboards/evolv/config.h:21]: Did you mean to leave this as "feed"?

**Review by drashna — APPROVED** [review]

File: keyboards/evolv/config.h (hunk 4 above)
Comment (Gondolindrim) [reply keyboards/evolv/config.h]: Updated for a better one in d1d1e82

File: keyboards/evolv/rules.mk [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+CONSOLE_ENABLE = yes        # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
```
Comment (fauxpark) [inline keyboards/evolv/rules.mk:19]:
```suggestion
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/evolv/rules.mk [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# MCU name
+MCU = STM32F072
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Virtual DIP switch configuration
```
Comment (fauxpark) [inline keyboards/evolv/rules.mk:7]:
```suggestion
BOOTMAGIC_ENABLE = full     # Virtual DIP switch configuration
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]
Missing an info.json, otherwise LGTM.

**Comment by Gondolindrim** [comment]
> Missing an info.json, otherwise LGTM.

Addded info.json

**Comment by fauxpark** [comment]
The info.json ordering doesn't seem to match the layout macro, for example the key after Insert is meant to be MPLY per the keymap, but the info.json has `KC_GRV` next:
image
You can test this by importing it into the Configurator with Ctrl+Shift+I, then turning on Fast Input and assigning keys.

**Comment by Gondolindrim** [comment]
> The info.json ordering doesn't seem to match the layout macro, for example the key after Insert is meant to be MPLY per the keymap, but the info.json has `KC_GRV` next:
> image
> You can test this by importing it into the Configurator with Ctrl+Shift+I, then turning on Fast Input and assigning keys.

Fixed in 3b2dbcb

**Comment by fauxpark** [comment]
Still doesn't seem correct:
image

**Comment by Gondolindrim** [comment]
> Still doesn't seem correct:
> image

Try ab62f1520

**Review by fauxpark — APPROVED** [review]

**Review by zvecr — APPROVED** [review]

**Review by zvecr — CHANGES_REQUESTED** [review]
Wait, i was going to hit merge, but the lint process is throwing errors with your info.json/layout macros

**Comment by Gondolindrim** [comment]
> Wait, i was going to hit merge, but the lint process is throwing errors with your info.json/layout macros

The problem was that despite the old info.json supporting only the ANSI version, the layout itself supports also ISO which has one more key (the backslash from the split left shift). 6e71068 introduces two layouts, `LAYOUT_evolv_iso` and `LAYOUT_evolv_ansi` which are both supported by the configurator now. The VIA layout naturally is the ISO layout in order to support all 88 keys. Lint checks now.

**Review by noroadsleft — COMMENTED** [review]
I imagine these aren't the correct layouts.

ISO:\
image

ANSI:\
image

Last column of keys is for the encoder, yes? And you four cardinal directions, and a click down? (Plus the spin actions, of course.)  Which function is which action?

**Comment by Gondolindrim** [comment]
> I imagine these aren't the correct layouts.
> 
> ISO:
> image
> 
> ANSI:
> image
> 
> Last column of keys is for the encoder, yes? And you four cardinal directions, and a click down? (Plus the spin actions, of course.) Which function is which action?

@noroadsleft I have no idea how to make the configurator json files. I just followed what is in the QMK documentation, and this has been done multiple times over. Is there any resource available showing precisely how the keys should be ordered in the file to make it correct on configurator?

**Comment by Gondolindrim** [comment]
@noroadsleft The latest commit rearranges the encoder keys to better picture the relation between movement and command. Also the `LAYOUT_evolv_ansi` and `LAYOUT_evolv_iso` were reworked to make it easier for the configurator. These should work fine now, tested with fast input.

**Pull request merged**